src/datasets/formatting/formatting.py

```diff
@@ -140,15 +185,39 @@ def extract_batch(self, pa_table: pa.Table) -> pa.Table:
         return pa_table
 
 
+def extract_struct_array(pa_array: pa.StructArray) -> list:
+    """StructArray.to_pylist / to_pydict does not call sub-arrays to_pylist / to_pydict methods so handle them manually."""
+    if isinstance(pa_array, pa.ChunkedArray):
+        batch_chunks = [extract_struct_array(chunk) for chunk in pa_array.chunks]
+        return [item for chunk in batch_chunks for item in chunk]
+
+    batch = {}
+    for field in pa_array.type:
+        if pa.types.is_struct(pa_array.field(field.name).type):
+            batch[field.name] = extract_struct_array(pa_array.field(field.name))
```

Comment:
I checked that lists of ArrayExtensionType features will call ArrayExtensionArray.to_pylist(), which didn't seem to be the case for struct.

Not sure about large list?

Reply:
cool ! maybe also check list of struct of ArrayExtensionType but no big deal, we can fix that rare case later (large list is also rare)

Reply:
the list of struct case might require an ArrayExtensionScalar or something with an as_py method that returns a numpy object.

Seems like it could be useful but have no idea whether this is possible or how best to do it if so?

Reply:
unless you know how to do this could we leave as issue?

Reply:
maybe just add a TODO comment about it for now ?